docs/drash/v2.x/2_tutorials/6_services/5_drash_approved_services/tengine.md

```diff
@@ -0,0 +1,160 @@
+
+## Folder Structure End State
+
+```text
+▾ /path/to/your/project/
+  ▾ views/
+    index.html
+    layout.html
+    user_details.html
+  app.ts
+  deps.ts
+```
+
+## Steps
+
+1. Create your `layout.html` template. This template will be extended by `index.html` via the `<% extends() %>` tag. When a template extends another template, the template that uses the `<% extends() %>` tag will be placed into the `<% yield %>` tag of the extended template.
+
+  ```html
+  <!DOCTYPE html>
+  <html class="h-full w-full">
+    <head>
+      <title>Drash + Tengine</title>
+      <link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/tailwindcss/dist/tailwind.min.css">
+    </head>
+    <body style="background: #f4f4f4">
+      <% yield %>
+    </body>
+  </html>
+  ```
+
+2. Create your `user_details.html` template partial. This will be included in `index.html`. As you can see, this template partial uses in-template JavaScript (`for` loop and `if/else` conditional). All in-template JavaScript will be evaluated by Tengine at runtime and rendered appropriately.
+
+  ```html
+  <ul>
+    <% for (let field in user.details) { %>
+      <!-- Do not show the Phone field -->
+      <% if (field !== "Phone") { %>
+    <li><% field %>: <% user.details[field] %></li>
+      <% } %>
+    <% } %>
+  </ul>
+  ```
+
+3. Create your `index.html` template. This template will extend `layout.html` via the `<% extends() %>` tag. This template will also include the `user_details.html` template partial via the `<% includes() %>` tag.
+
+  ```html
+  <% extends("/layout.html") %>
+
+  <div style="max-width: 640px; margin: 50px auto;">
+    <h1 class="text-5xl mb-5"><% user.name %></h1>
+    <% include_partial("/user_details.html") %>
+  </div>
+  ```
+
+4. Create your `app.ts` file.
+
+  ```typescript
+  // app.ts
+
+  import {
+    Drash,
+    TengineService,
+  } from "./deps.ts";
+
+  // Instantiate and configure TengineService
+
+  const tengine = new TengineService({
+    views_path: "./views",
+  });
+
+  // Create your resource with TengineService enabled on GET requests
+
+  class HomeResource extends Drash.Resource {
+    public paths = ["/"];
+
+    public GET(_request: Drash.Request, response: Drash.Response): void {
+      const templateVariables = {
+        user: {
+          name: "Jae",
+          details: {
+            "Role": "Software Engineer",
+            "Phone": "(555) 555-5555",
+            "E-mail": "jae@example.com",
+          },
+        }
+      };
+
+      const html = response.render("/index.html", templateVariables) as string;
+
+      response.html(html);
+    }
+  }
+
+  // Create and run your server
+
+  const server = new Drash.Server({
+    hostname: "0.0.0.0",
+    port: 1447,
+    protocol: "http",
+    resources: [HomeResource],
+    services: [tengine],
+  });
+
+  server.run();
+
+  console.log(`Server running at ${server.address}.`);
+  ```
+
+## Verification
+
+1. Run your app.
+
+  ```shell
+  $ deno run --allow-net app.ts
```

Comment:
Haven’t tested but won’t we need read perms too?

Reply:
good catch. yea. copypasta

Reply:
e706467